### ISSUE TITLE
cmd/go: go generate sets more variables than its documentation specifies

### ISSUE THREAD

**Issue by dmitshur** [golang/go#41943]
`go generate` [documentation](https://golang.org/cmd/go/#hdr-Generate_Go_files_by_processing_source) currently says:

> Go generate sets several variables when it runs the generator:
>
> ```
> $GOARCH
> 	The execution architecture (arm, amd64, etc.)
> $GOOS
> 	The execution operating system (linux, windows, etc.)
> $GOFILE
> 	The base name of the file.
> $GOLINE
> 	The line number of the directive in the source file.
> $GOPACKAGE
> 	The name of the package of the file containing the directive.
> $DOLLAR
> 	A dollar sign.
> ```
>
> [...] **If the variable NAME is not set, $NAME expands to the empty string.**

The last sentence is important. It leads the user to believe that if a NAME is not one of the aforementioned variables and is not set in the environment, $NAME should expand to the empty string.

This is not the case for some Go-specific variables (beyond those mentioned in the list above) because [`expandVar`](https://github.com/golang/go/blob/8994607f828e46272bae5e7959997dea8cb1a19a/src/cmd/go/internal/generate/generate.go#L418-L428) uses `os.Getenv(word)` to look up environment variables that aren't in `g.env`. To achieve the behavior as documented above, it should instead be looking up environment variables from [`cfg.OrigEnv`](https://golang.org/cmd/go/internal/cfg/#OrigEnv), since cmd/go's `main` modifies the environment at startup (see [here](https://github.com/golang/go/blob/8994607f828e46272bae5e7959997dea8cb1a19a/src/cmd/go/main.go#L146-L157)).

To reproduce, consider this program:

```Go
package main

import (
	"fmt"
	"os"
)

//go:generate echo ${DOLLAR}GOCACHE expands to '$GOCACHE'

func main() {
	v, ok := os.LookupEnv("GOCACHE")
	if !ok {
		fmt.Println("env var GOCACHE is unset")
		return
	}
	fmt.Printf("$GOCACHE = %q\n", v)
}
```

### What did you do?

```
$ unset GOCACHE
$ go run .
$ go generate
```

### What did you expect to see?

```
$ go run .
env var GOCACHE is unset
$ go generate
$GOCACHE expands to ''
```

### What did you see instead?

```
$ go run .
env var GOCACHE is unset
$ go generate
$GOCACHE expands to '/Users/gopher/Library/Caches/go-build'
```

/cc @bcmills @jayconrod @matloob

**Comment by dmitshur** [comment]
NeedsDecision is to decide whether to fix the implementation or the documentation.

**Comment by mvdan** [comment]
I think the current documentation is generally correct: it says it *sets* some variables, it doesn't say anything about starting from an empty environment. So, one would expect that environment variables from the parent process would be set, like `HOME`.

Perhaps we can clarify the documentation to say which of the `go env` variables are always present too - presumably all of them.

**Comment by dmitshur** [comment]
> it doesn't say anything about starting from an empty environment

I agree. This issue isn't expecting an empty environment plus the 6 documented variables. It's expecting the caller's environment plus the 6 documented variables (what the documentation suggests).

The current implementation produces the caller's environment plus a subset of `go env` (undocumented) plus 6 documented variables.

**Comment by mvdan** [comment]
Right, apologies for misreading. Ignore the first part of my previous comment. I still think that this just needs clearer documentation, and I think that we should expose all of `go env` for consistency.
